crates/bevy_ui_widgets/src/radio.rs

```diff
@@ -141,82 +152,101 @@ fn radio_group_on_key_input(
+// on `Space` or `Enter` key press.
+fn radio_button_on_key_input(
+    mut ev: On<FocusedInput<KeyboardInput>>,
+    q_radio_button: Query<Has<Checked>, (With<RadioButton>, Without<InteractionDisabled>)>,
     q_group: Query<(), With<RadioGroup>>,
-    q_radio: Query<(Has<Checked>, Has<InteractionDisabled>), With<RadioButton>>,
     q_parents: Query<&ChildOf>,
-    q_children: Query<&Children>,
     mut commands: Commands,
 ) {
-    if q_group.contains(ev.entity) {
-        // Starting with the original target, search upward for a radio button.
-        let radio_id = if q_radio.contains(ev.original_event_target()) {
-            ev.original_event_target()
-        } else {
-            // Search ancestors for the first radio button
-            let mut found_radio = None;
-            for ancestor in q_parents.iter_ancestors(ev.original_event_target()) {
-                if q_group.contains(ancestor) {
-                    // We reached a radio group before finding a radio button, bail out
-                    return;
-                }
-                if q_radio.contains(ancestor) {
-                    found_radio = Some(ancestor);
-                    break;
-                }
-            }
+    let Ok(checked) = q_radio_button.get(ev.focused_entity) else {
+        // Not a radio button
+        return;
+    };
+
+    // Radio button already checked
+    if checked {
+        return;
+    }
 
-            match found_radio {
-                Some(radio) => radio,
-                None => return, // No radio button found in the ancestor chain
-            }
-        };
+    let event = &ev.event().input;
+    if event.state == ButtonState::Pressed
+        && !event.repeat
+        && (event.key_code == KeyCode::Enter || event.key_code == KeyCode::Space)
+    {
+        ev.propagate(false);
 
-        // Radio button is disabled.
-        if q_radio.get(radio_id).unwrap().1 {
-            return;
-        }
+        trigger_radio_button_and_radio_group_value_change(
+            ev.focused_entity,
+            &q_group,
+            &q_parents,
+            &mut commands,
+        );
+    }
+}
 
-        // Gather all the enabled radio group descendants for exclusion.
-        let radio_buttons = q_children
-            .iter_descendants(ev.entity)
-            .filter_map(|child_id| match q_radio.get(child_id) {
-                Ok((checked, false)) => Some((child_id, checked)),
-                Ok((_, true)) | Err(_) => None,
-            })
-            .collect::<Vec<_>>();
-
-        if radio_buttons.is_empty() {
-            return; // No enabled radio buttons in the group
-        }
+fn radio_button_on_click(
+    mut ev: On<Pointer<Click>>,
+    q_group: Query<(), With<RadioGroup>>,
+    q_radio: Query<Has<Checked>, (With<RadioButton>, Without<InteractionDisabled>)>,
+    q_parents: Query<&ChildOf>,
+    mut commands: Commands,
+) {
+    let Ok(checked) = q_radio.get(ev.entity) else {
```

Comment:
Bailing out here means that a disabled radio button doesn't consume the click event, which will propagate to the ancestors. In other words, it will cause the click to trigger on whatever container entity holds the radio button, which I think is probably incorrect.

I think that disabled widgets shouldn't behave as if the widget was absent; rather, they should behave as if the widget was read-only. This means that they still receive click events, they just don't respond to them.

This is why I used `Has<InteractionDisabled>` rather than `Without<InteractionDisabled>` - so that I could detect clicks on disabled buttons and stop propagation on them.

Reply:
I tried to implement behaviour based on the current click implementation for radio group / button.

Current implementation doesn't  cancel click event propagation on a disabled widget.

https://github.com/bevyengine/bevy/blob/eac8bc7cc4b35c98db00a6cf1fd39377afcbac61/crates/bevy_ui_widgets/src/radio.rs#L176-L196

I can fix that too in this PR.